src/test/connect-test.c

```diff
@@ -0,0 +1,444 @@
+        }
+
+        r = setsockopt(listen_fd_6, SOL_SOCKET, SO_REUSEADDR, &reuseaddr, sizeof(reuseaddr));
+        if (r < 0)
+                goto end;
+
+        serv_addr_6.sin6_family = AF_INET6;
+        serv_addr_6.sin6_port = htons(port);
+        r = inet_pton(AF_INET6, "::1", &serv_addr_6.sin6_addr);
+        if (r < 0)
+                goto end;
+
+        r = bind(listen_fd_6, (struct sockaddr *) &serv_addr_6, sizeof(serv_addr_6));
+        if (r < 0)
+                goto end;
+
+        r = listen(listen_fd_6, 1);
+        if (r < 0)
+                goto end;
+
+        r = sd_event_add_io(e, &event_source, listen_fd_6, EPOLLIN, server_handler_6, &ctx);
+        if (r < 0)
+                goto end;
+
+        /* client sockets */
+
+        fd_4 = socket(AF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
+        if (fd_4 < 0) {
+                r = -errno;
+                goto end;
+        }
+
+        r = setsockopt(fd_4, SOL_SOCKET, SO_REUSEADDR, &reuseaddr, sizeof(reuseaddr));
+        if (r < 0)
+                goto end;
+
+        /* only connect to localhost */
+
+        /*
+         * Bind client socket to port+1. This allows for egress testing, because
+         * we know which port the server will connect back.
+         */
+
+        addr_4.sin_family = AF_INET;
+        addr_4.sin_port = htons(port+1);
+        r = inet_pton(AF_INET, "127.0.0.1", &addr_4.sin_addr);
+        if (r < 0)
+                goto end;
+
+        r = bind(fd_4, (struct sockaddr *) &addr_4, sizeof(addr_4));
+        if (r < 0) {
+                fprintf(stderr, "IPv4 client bind failed: %s\n", strerror(errno));
+                goto end;
+        }
+
+        addr_4.sin_port = htons(port);
+
+        /* try to connect both sockets */
+        r = connect(fd_4, (struct sockaddr *) &addr_4, sizeof(addr_4));
+        if (r == 0)
+                ctx.ipv4 = CONNECTION_SUCCEEDED;
+        else if (r < 0 && errno == EINPROGRESS) {
+                r = sd_event_add_io(e, &event_source, fd_4, EPOLLOUT, conn_handler_4, &ctx);
+                if (r < 0)
+                        goto end;
+        } else {
+                fprintf(stderr, "IPv4 connection failed: %s\n", strerror(r));
+                goto end;
+        }
+
+        fd_6 = socket(AF_INET6, SOCK_STREAM | SOCK_NONBLOCK, 0);
+        if (fd_6 < 0) {
+                r = -errno;
+                goto end;
+        }
+
+        r = setsockopt(fd_6, SOL_SOCKET, SO_REUSEADDR, &reuseaddr, sizeof(reuseaddr));
+        if (r < 0)
+                goto end;
+
+        addr_6.sin6_family = AF_INET6;
+        addr_6.sin6_port = htons(port+1);
+        r = inet_pton(AF_INET6, "::1", &addr_6.sin6_addr);
+        if (r < 0)
+                goto end;
+
+        r = bind(fd_6, (struct sockaddr *) &addr_6, sizeof(addr_6));
+        if (r < 0) {
+                fprintf(stderr, "IPv6 client bind failed: %s\n", strerror(r));
+                goto end;
+        }
+
+        addr_6.sin6_port = htons(port);
```

Comment:
we generaly avoid htons() and friends, and use htobe16() and friends instead as it is mor descriptive and correct